### ISSUE TITLE
Puzzling problem with constrint

### ISSUE THREAD

**Issue by tomjaguarpaw** [haskell/error-messages#27]
I was trying to implement an [example](https://hackage.haskell.org/package/tagsoup-0.14.8/docs/Text-HTML-TagSoup.html#t:ParseOptions) from the tagsoup package.  (The example has typos but that's not relevant here.)

Specifically, there is a value

```haskell
parseOptions :: StringLike str => ParseOptions str 
```

and I want to override some of its fields at `str = String`

```haskell
parseOptionsPreserveEntities :: ParseOptions String
parseOptionsPreserveEntities = parseOptions
  { optEntityData = \(str,b) -> [TagText $ "&" ++ str ++ [';' | b]]
  , optEntityAttrib = \(str,b) -> ("&" ++ str ++ [';' | b], [])
  }
```

But this doesn't work

```
app/Main.hs:10:32: error:
    • Ambiguous type variable ‘str0’ arising from a use of ‘parseOptions’
      prevents the constraint ‘(Text.StringLike.StringLike
                                  str0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘str0’ should be.
      These potential instances exist:
        instance [safe] Text.StringLike.StringLike String
          -- Defined in ‘Text.StringLike’
        ...plus four instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    • In the expression: parseOptions
      In the expression:
        parseOptions
          {optEntityData = \ (str, b) -> [TagText $ "&" ++ str ++ ...],
           optEntityAttrib = \ (str, b) -> ("&" ++ str ++ [';' | b], [])}
      In an equation for ‘parseOptionsPreserveEntities’:
          parseOptionsPreserveEntities
            = parseOptions
                {optEntityData = \ (str, b) -> [...],
                 optEntityAttrib = \ (str, b) -> ("&" ++ str ++ ..., [])}
   |
10 | parseOptionsPreserveEntities = parseOptions
   |                                ^^^^^^^^^^^^...
```

I don't actually understand *why* it doesn't work. A workaround is to convert it into two record updates:

```haskell
parseOptionsPreserveEntities :: ParseOptions String
parseOptionsPreserveEntities = parseOptions
  { optEntityData = \(str,b) -> [TagText $ "&" ++ str ++ [';' | b]] }
  { optEntityAttrib = \(str,b) -> ("&" ++ str ++ [';' | b], []) }
```

I'm not sure whether I can suggest a candidate for a good error message because I don't actually understand what's going on. Perhaps multiple record updates desugar into something that breaks type inference. Perhaps this could be detected and a helpful suggestion to split the update into multiple single-field updates could be emitted.

(GHC 8.10.7)

**Comment by adamgundry** [comment]
This is arising as a result of type-changing record updates. The problem is that in the occurrence of `parseOptions :: StringLike str => ParseOptions str` prior to the update, the `str` type variable is genuinely ambiguous, as the message states.

When you update both the `optEntityData` and `optEntityAttrib` fields at once, you can have different parameter types for the record before and after the update, because `str` occurs only in the types of those two fields in the definition of `ParseOptions`. The type of the parameter after the update is fixed to be `String` by the type signature and field values, but there is nothing to determine what the type of the record before the update should be.

When you have two separate record updates, the record type cannot change, because neither update sets *all* the fields that reference the `str` parameter. Thus there is no ambiguity and `String` is used throughout.

Even knowing this I'm not sure how to make the error message clearer, alas.

**Comment by Ericson2314** [comment]
Why is

> Ambiguous type variable ... prevents the constraint ... from being solved

an error at all? Is the variable not just as ambiguous and problematic without the constraint? Can we get GHC to do bind the type variable, be happy with the constraint, and *then* complain about ambiguity?

I rather not drag irrelevant extra details into the errors like this if possible.

**Comment by adamgundry** [comment]
> Is the variable not just as ambiguous and problematic without the constraint?

No, it isn't. Ambiguous unconstrained type variables aren't a problem because no runtime behaviour depends on their value, so it doesn't matter how they are instantiated. (In practice GHC defaults them to a magic built-in type family called `Any`.) For example the expression `null []` is accepted even though the list type is ambiguous.

**Comment by Ericson2314** [comment]
Ah OK, thanks for cluing me in. I guess I had the pre-`Any` model in my head?

Anyways, what makes this and #7 have different sorts of error messages? (In general, GHC seems it has more variations on error messages than errors, which has left me feeling confused at times.)

**Comment by adamgundry** [comment]
> Anyways, what makes this and #7 have different sorts of error messages?

That's a very good question. I don't know for sure what makes GHC sometimes issue "Ambiguous type variable ... arising from a use of ... prevents the constraint ... from being solved." and at other times "Could not deduce ...  arising from a use of ... 
 from the context ... The type variable ... is ambiguous.". At a guess, perhaps the former is used when there are no given constraints, while the latter is used if there are some?

It would be simpler and more consistent to have a single message form in both cases. For the error in the issue description, we could imagine using the message from #7 but omitting the "from the context" stanza(s):
```
Could not deduce (Text.StringLike.StringLike str0)
  arising from a use of ‘parseOptions’
The type variable ‘str0’ is ambiguous
```

That said, there is something to be said for flagging up the ambiguous type variable explicitly as the likely source of the problem. So one could also go the other way and turn the message in #7 into this?
```
Ambiguous type variable ‘m0’ arising from a use of ‘easyRender’
prevents the constraint
    (ToGVal (Run SourcePos (Control.Monad.Trans.Writer.Lazy.Writer Text) Text) (GVal m0))
from being solved
   from the context: ToJSON c
        bound by the type signature for:
                   renderTemplate :: forall c.
                                     ToJSON c =>
                                     Text -> c -> Either Text Text
        at Ginger.hs:14:1-59

**Comment by Ericson2314** [comment]
OK glad it's not just me! Yes, I would like putting the "ambiguous variable" part first. I have felt like the #7 one buries the lede, whenever I was confronted with it.

**Comment by tomjaguarpaw** [comment]
> When you have two separate record updates, the record type cannot change, because neither update sets all the fields that reference the `str` parameter.

Oh wow, that is subtle!

**Comment by goldfirere** [comment]
Yes, I think it's the presence of Givens that distinguishes the messages, and I agree this is confusing. I personally think the ambiguous variable should lead the charge.
